akka-remote/src/test/scala/akka/remote/ActorsLeakSpec.scala

```diff
@@ -0,0 +1,176 @@
+package akka.remote
+
+import java.util.concurrent.TimeoutException
+
+import akka.actor._
+import akka.actor.dungeon.ChildrenContainer
+import akka.event.Logging
+import akka.remote.transport.ThrottlerTransportAdapter.Direction.Both
+import akka.remote.transport.ThrottlerTransportAdapter.{ ForceDisassociate, Blackhole, SetThrottle }
+import akka.testkit._
+import akka.testkit.TestActors.EchoActor
+import com.typesafe.config.ConfigFactory
+
+import scala.collection.immutable
+import scala.concurrent.Await
+import scala.concurrent.duration._
+
+object ActorsLeakSpec {
+
+  val config = ConfigFactory.parseString(
+    """
+      | akka.actor.provider = "akka.remote.RemoteActorRefProvider"
+      | #akka.loglevel = DEBUG
+      | akka.remote.netty.tcp.applied-adapters = ["trttl"]
+      | #akka.remote.log-lifecycle-events = on
+      | akka.remote.transport-failure-detector.heartbeat-interval = 1 s
+      | akka.remote.transport-failure-detector.acceptable-heartbeat-pause = 3 s
+      | akka.remote.quarantine-after-silence = 3 s
+      | akka.test.filter-leeway = 10 s
+      |
+      |""".stripMargin)
+
+  def collectLiveActors(root: ActorRef): immutable.Seq[ActorRef] = {
+
+    def recurse(node: ActorRef): List[ActorRef] = {
+      val children: List[ActorRef] = node match {
+        case wc: ActorRefWithCell ⇒
+          val cell = wc.underlying
+
+          cell.childrenRefs match {
+            case ChildrenContainer.TerminatingChildrenContainer(_, toDie, reason) ⇒ Nil
+            case x @ (ChildrenContainer.TerminatedChildrenContainer | ChildrenContainer.EmptyChildrenContainer) ⇒ Nil
+            case n: ChildrenContainer.NormalChildrenContainer ⇒ cell.childrenRefs.children.toList
+            case x ⇒ Nil
+          }
+        case _ ⇒ Nil
+      }
+
+      node :: children.flatMap(recurse)
+    }
+
+    recurse(root)
+  }
+
+  class StoppableActor extends Actor {
+    override def receive = {
+      case "stop" ⇒ context.stop(self)
+    }
+  }
+
+}
+
+class ActorsLeakSpec extends AkkaSpec(ActorsLeakSpec.config) with ImplicitSender {
+  import ActorsLeakSpec._
+
+  "Remoting" must {
+
+    "not leak actors" in {
+      val ref = system.actorOf(Props[EchoActor], "echo")
+      val echoPath = RootActorPath(RARP(system).provider.getDefaultAddress) / "user" / "echo"
+
+      val initialActors =
+        (collectLiveActors(system.actorFor("/system/endpointManager")) ++
+          collectLiveActors(system.actorFor("/system/transports"))).toSet
+
+      //Clean shutdown case
+      for (_ ← 1 to 3) {
+
+        val remoteSystem = ActorSystem(
+          "remote",
+          ConfigFactory.parseString("akka.remote.netty.tcp.port = 2553")
```

Comment:
don't use hardcoded ports, since that prevents parallel execution on same machine (when we need that)